packages/slate/src/classes/commands/node.ts

```diff
@@ -0,0 +1,773 @@
+
+      const toRef = this.createPathRef(to)
+      const targets = this.matches({ at, match })
+      const pathRefs = Array.from(targets, ([, p]) => this.createPathRef(p))
+
+      for (const pathRef of pathRefs) {
+        const path = pathRef.unref()!
+        const newPath = toRef.current!
+
+        if (path.length !== 0) {
+          this.apply({ type: 'move_node', path, newPath })
+        }
+      }
+
+      toRef.unref()
+    })
+  }
+
+  /**
+   * Normalize a node at a path, returning it to a valid state if it is
+   * currently invalid.
+   */
+
+  normalizeNodes(this: Editor, options: { at: Path }): void {
+    const { at } = options
+    const [node] = this.getNode(at)
+
+    // There are no core normalizations for text nodes.
+    if (Text.isText(node)) {
+      return
+    }
+
+    // Ensure that block and inline nodes have at least one text child.
+    if (Element.isElement(node) && node.nodes.length === 0) {
+      const child = { text: '', marks: [] }
+      this.insertNodes(child, { at: at.concat(0) })
+      return
+    }
+
+    // Determine whether the node should have block or inline children.
+    const shouldHaveInlines =
+      Element.isElement(node) &&
+      (this.isInline(node) ||
+        node.nodes.length === 0 ||
+        Text.isText(node.nodes[0]) ||
+        this.isInline(node.nodes[0]))
+
+    // Since we'll be applying operations while iterating, keep track of an
+    // index that accounts for any added/removed nodes.
+    let n = 0
+
+    for (let i = 0; i < node.nodes.length; i++, n++) {
+      const child = node.nodes[i] as Descendant
+      const prev = node.nodes[i - 1]
+      const isLast = i === node.nodes.length - 1
+
+      if (Element.isElement(child)) {
+        const isInline = this.isInline(child)
+
+        // Only allow block nodes in the top-level value and parent blocks that
+        // only contain block nodes. Similarly, only allow inline nodes in other
+        // inline nodes, or parent blocks that only contain inlines and text.
+        if (isInline !== shouldHaveInlines) {
+          this.removeNodes({ at: at.concat(n) })
+          n--
+          continue
+        }
+
+        // Ensure that inline nodes are surrounded by text nodes.
+        if (isInline) {
+          if (prev == null || !Text.isText(prev)) {
+            const child = { text: '', marks: [] }
+            this.insertNodes(child, { at: at.concat(n) })
+            n++
+            continue
+          }
+
+          if (isLast) {
+            const child = { text: '', marks: [] }
+            this.insertNodes(child, { at: at.concat(n + 1) })
+            n++
+            continue
+          }
+        }
+      } else {
+        // Merge adjacent text nodes that are empty or have matching marks.
+        if (prev != null && Text.isText(prev)) {
+          if (Text.matches(child, prev)) {
+            this.mergeNodes({ at: at.concat(n) })
+            n--
+            continue
+          } else if (prev.text === '') {
```

Comment:
Should these two `else if` conditions check to make sure one of the nodes does not have different marks than the other before merging? If I create a new empty text node and add a mark, then this merges it with the previous text node.

Reply:
Good question, I think it should you're right.

Reply:
Although, I'm not sure extra empty text nodes should be allowed actually. So maybe it shouldn't.